src/lib/backend/equinoxe_cohttp.ml

```diff
@@ -0,0 +1,129 @@
+open Lwt.Infix
+module Client = Cohttp_lwt_unix.Client
+module Json = Equinoxe.Json
+
+module Backend = struct
+  (**** Type definitions ****)
+
+  type t = { address : string; token : string }
+
+  let token t = t.token
+  let address t = t.address
+
+  (**** Default values ****)
+
+  let build_header token =
+    let token = if token = "" then [] else [ ("X-Auth-Token", token) ] in
+    token @ [ ("Content-Type", "application/json") ] |> Cohttp.Header.of_list
+
+  (***** Helpers *****)
+
+  let compute ~time ~f =
+    let* status =
+      Lwt.pick
+        [
+          (f () >|= fun v -> `Done v);
+          (Lwt_unix.sleep time >|= fun () -> `Timeout);
+        ]
+    in
+    match status with
+    | `Done v -> Lwt.return v
+    | `Timeout -> Lwt.fail_with "Http request timeout"
+
+  let convert_to_json (resp, body) =
+    let code = Cohttp.(Response.status resp |> Code.code_of_status) in
+    if code >= 200 && code < 300 then
+      Lwt.catch
+        (fun () ->
+          let+ body = Cohttp_lwt.Body.to_string body in
+          match body with "" -> Json.of_string "{ }" | s -> Json.of_string s)
+        (fun e -> Lwt.return @@ Json.error (Printexc.to_string e))
+    else
+      let msg = Format.sprintf "Cohttp exits with HTTP code %d" code in
+      Lwt.return @@ Json.error msg
+
+  (**** Http methode ****)
+
+  let compute = compute ~time:10.0
+
+  let get_from t path =
+    let headers = build_header t.token in
+    let url = Filename.concat t.address path |> Uri.of_string in
+    let f () = Client.get ~headers url in
+    compute ~f
+
+  let post_from t ~path body =
+    let headers = build_header t.token in
+    let url = Filename.concat t.address path |> Uri.of_string in
+    let body = Cohttp_lwt.Body.of_string body in
+    let f () = Client.post ~headers ~body url in
+    compute ~f
+
+  let put_from t ~path body =
+    let headers = build_header t.token in
+    let url = Filename.concat t.address path |> Uri.of_string in
+    let body = Cohttp_lwt.Body.of_string body in
+    let f () = Client.put ~headers ~body url in
+    compute ~f
+
+  let delete_from t path =
+    let headers = build_header t.token in
+    let url = Filename.concat t.address path |> Uri.of_string in
+    let f () = Client.delete ~headers url in
+    compute ~f
+
+  (**** API ****)
+
+  let create ~address ~token () = { address; token }
+
+  let get t ~path () =
+    let* resp = get_from t path in
+    convert_to_json resp
+
+  let post t ~path json =
+    match Json.export json with
+    | Ok body ->
+        let* resp = post_from t ~path body in
+        convert_to_json resp
+    | Error (`Msg e) -> Lwt.return @@ Json.error e
+
+  let put t ~path json =
+    match Json.export json with
+    | Ok body ->
+        let* resp = put_from t ~path body in
+        convert_to_json resp
+    | Error (`Msg e) -> Lwt.return @@ Json.error e
+
+  let delete t ~path () =
+    let* resp = delete_from t path in
+    convert_to_json resp
+
+  let run json = Lwt_main.run json
```

Comment:
You probably don't want to call this `run` on all the http calls, rather the Equinoxe api should also produce Lwt values when the backend requires it. To stay compatible with "direct style" I/O, you can abstract the monad in the backend:

```ocaml
type 'a m
val return : 'a -> 'a m
val bind : ('a -> 'b m) -> 'a m -> 'b m
(* ... *)
val post : t -> path:string -> Json.t -> Json.t m
```

Then the backend can instantiate with `type 'a m = 'a Lwt.t` or `type 'a m = 'a` (here's an [example](https://github.com/art-w/ocaml-webdriver/blob/master/src/webdriver_sig.ml) of what it might look like)